### PR TITLE
implement-autoScroll

### LINKED ISSUE #137: Automatically scroll down on students page text area when new message arrives
Step 1: Automatically scroll down on students page text area when new message arrives

Step 2: we will place a checkbox to disable that behavior

### PR DESCRIPTION
**Pull request by manudous**
Closed #137 @brauliodiez

### CODE REVIEW

File: front/src/pods/student/student.component.tsx [hunk 1]
```diff
@@ -11,25 +16,44 @@ interface Props {
 export const StudentComponent: React.FC<Props> = props => {
   const { room, log } = props;
 
+  const [autoScroll, setAutoScroll] = React.useState(false);
```
Comment (brauliodiez) [inline front/src/pods/student/student.component.tsx:19]: Let's rename this to _isAutoScrollEnabled_ and _setIsAutoScrollEnabled_

```tsx
const [isAutoScrollEnabled, setAutoScrollEnabled] = React.useState(false);
```

By doing this the code is more readable, you kow exactly what does holds this state

Comment (brauliodiez) [reply front/src/pods/student/student.component.tsx:19]: By default it should be true

File: front/src/pods/student/student.component.tsx [hunk 2]
```diff
@@ -11,25 +16,44 @@ interface Props {
...
+
+  const textAreaRef = React.useRef(null);
+
+  React.useEffect(() => {
+    if (!autoScroll)
+      textAreaRef.current.scrollTop = textAreaRef.current.scrollHeight;
```
Comment (brauliodiez) [inline front/src/pods/student/student.component.tsx:25]: Taking the update form the previous it  you don't need the not operator, and on the other hand let's apply some defensive programming, _textAreaRef_ will always hold a value, but image that somebody decides to hide the control temporary by some reason, rather avoid a bad crash here.

```tsx
if (isAutoScrollEnabled && textAreaRef.current) {
      textAreaRef.current.scrollTop = textAreaRef.current.scrollHeight;
}
```

On important thing always add the curly bracket to an if statement, even if it's one line

https://stackoverflow.com/questions/2125066/is-it-a-bad-practice-to-use-an-if-statement-without-curly-braces

File: front/src/pods/trainer/components/session.component.tsx [hunk 3]
```diff
@@ -30,8 +34,14 @@ const getFullContent = (currenSessionContent: string) => {
 export const SessionComponent: React.FC<Props> = props => {
   const { log, handleSendFullContentLog, className } = props;
 
+  const [autoScroll, setAutoScroll] = React.useState(false);
```
Comment (brauliodiez) [inline front/src/pods/trainer/components/session.component.tsx:37]: Same changes as in previous file.
And once you are done we will extract this to a generic hook 😇

**Comment by manudous** [comment]
Now time to the Hook xd

File: front/src/pods/student/student.component.tsx [hunk 4]
```diff
@@ -11,25 +15,48 @@ interface Props {
...
 
+  const [isAutoScrollEnabled, setAutoScrollEnabled] = React.useState(true);
+
+  const textAreaRef = React.useRef(null);
+
+  React.useEffect(() => {
+    if (isAutoScrollEnabled && textAreaRef.current) {
+      textAreaRef.current.scrollTop = textAreaRef.current.scrollHeight;
+    }
+  }, [log]);
```
Comment (brauliodiez) [inline front/src/pods/student/student.component.tsx:26]: Ok, Let's try to create reusable effect, psuedocode:

Option A:

```tsx
const useAutoScroll = () => {
  const [isAutoScrollEnabled, setIsAutoScrollEnabled] = React.useState(true);
  const textAreaRef = React.useRef(null);

  const doAutoScroll = () => {
     if(isAutoScrollEnabled && txtAreaRef.current) {
      textAreaRef.current.scrollTop = textAreaRef.current.scrollHeight;
    }
  }

  return {isAutoScrollEnabled, setIsAutoScrollEnabled, textAreaRef, doAutoScroll}
}
```

Usage

```tsx
const {isAutoScrollEnabled, setIsAutoScrollEnabled, textAreaRef, doAutoScroll} = useAutoScroll();

React.useEffect(() => {
  doAutoScroll();
}, [log]);
```

(...)

Not sure if we could some how pass 'log' (I think is not a good idea to keep it in the effect), and maybe we could create a textAreaWrapper that includes this funcionallity but maybe is just an overkill, I think just with this is good enough

Comment (manudous) [reply front/src/pods/student/student.component.tsx:26]: Ok, I was thinking of doing the same, where should I keep the hook?

File: front/src/pods/student/student.component.tsx [hunk 5]
```diff
@@ -1,7 +1,12 @@
 import React from 'react';
 import TextareaAutosize from '@material-ui/core/TextareaAutosize';
 import Typography from '@material-ui/core/Typography';
+
+import Checkbox from '@material-ui/core/Checkbox';
+import FormControlLabel from '@material-ui/core/FormControlLabel';
+
 import * as innerClasses from './student.styles';
+import { useAutoScroll } from '../../common/hooks/use-Auto-Scroll';
```
Comment (Nasdan) [inline front/src/pods/student/student.component.tsx:9]: The `common`  folder has an alias. That means you can write the import like:

`import { useAutoScroll } from 'common/hooks/use-Auto-Scroll';`

File: front/src/common/hooks/use-Auto-Scroll.tsx [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
```
Comment (Nasdan) [inline front/src/common/hooks/use-Auto-Scroll.tsx:1]: It's better to name files with lowercase. In this case, I usually name this file like: `auto-scroll.hook.ts`

File: front/src/common/hooks/auto-scroll.hook.tsx [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+export const useAutoScroll = () => {
```
Comment (brauliodiez) [inline front/src/common/hooks/auto-scroll.hook.tsx:3]: Missing auto-scroll.hook.spect.ts :)


We can give a try tomorrow :)

**Pull request merged**